### ISSUE TITLE
Hints not triggered in typeclass instances

### ISSUE THREAD

**Issue by zliu41** [ndmitchell/hlint#839]
The culprit appears to be [this line](https://github.com/ndmitchell/hlint/blob/f404c85ef5e14bff6e0216072e04ea040821fbbd/src/Hint/Match.hs#L115). `HsDecl`, as far as I can tell, is not recursive (unlike `Decl`), so `children` always returns `[]`.

But we can't simply delete that line, because `checkDefine'` won't work properly and [this test](https://github.com/ndmitchell/hlint/blob/f404c85ef5e14bff6e0216072e04ea040821fbbd/data/hlint.yaml#L918) will break.

A simple but hacky solution is to add an `InstD` case to `declName`, returning a list of names, and if any name matches `y` in `checkDefine'`, we pass.

A less hacky solution is to update `universeParentExp'` to additionally return the `fun_id` and `var_id` associated with each `LHsExpr` (if exists), which are then passed to `checkDefine'` for comparison.

Happy to make a PR for either approach (or a different one).

**Comment by ndmitchell** [comment]
Thanks for the report. I came up with a small test case:

```
instance Class X where
    method = map f (map g x)
```

I confirm it's broken. In trying out the solutions you described to figure out which one worked best, I ended up solving it, so I've committed. I think it's basically the second solution. Much appreciated for your offer of a fix though and figuring out how to fix it!
